<!DOCTYPE html>
<html>

  <head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  <title>Basic Command line tools for bioinformatics</title>
  <meta name="description" content="So I’m in week 02 of the Coursera course  Command Line Tools for Genomic Data Science  and I am really having fun.I feel like I learned a lot already. It’s b...">

  <link rel="stylesheet" href="/digital-codon/css/main.css">
  <link rel="canonical" href="https://nezlicodes.io/digital-codon/bioinfomatics/genomic/data/science/2020/04/15/Basic-Command-line-tools-for-bioinformatics.html">
  <link rel="alternate" type="application/rss+xml" title="Digital codon" href="https://nezlicodes.io/digital-codon/feed.xml">
</head>


  <body>

    <header class="site-header">

  <div class="wrapper">

    <a class="site-title" href="/digital-codon/">Digital codon</a>

    <nav class="site-nav">
      <a href="#" class="menu-icon">
        <svg viewBox="0 0 18 15">
          <path fill="#424242" d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.031C17.335,0,18,0.665,18,1.484L18,1.484z"/>
          <path fill="#424242" d="M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0c0-0.82,0.665-1.484,1.484-1.484 h15.031C17.335,6.031,18,6.696,18,7.516L18,7.516z"/>
          <path fill="#424242" d="M18,13.516C18,14.335,17.335,15,16.516,15H1.484C0.665,15,0,14.335,0,13.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.031C17.335,12.031,18,12.696,18,13.516L18,13.516z"/>
        </svg>
      </a>

      <div class="trigger">
        
          
          <a class="page-link" href="/digital-codon/about/">About</a>
          
        
          
        
          
        
          
        
      </div>
    </nav>

  </div>

</header>


    <div class="page-content">
      <div class="wrapper">
        <article class="post" itemscope itemtype="http://schema.org/BlogPosting">

  <header class="post-header">
    <h1 class="post-title" itemprop="name headline">Basic Command line tools for bioinformatics</h1>
    <p class="post-meta"><time datetime="2020-04-15T16:58:59+02:00" itemprop="datePublished">Apr 15, 2020</time></p>
  </header>

  <div class="post-content" itemprop="articleBody">
    <p>So I’m in week 02 of the Coursera course <a href="https://www.coursera.org/learn/genomic-tools" target="_about"> Command Line Tools for Genomic Data Science </a> and I am really having fun.</p>

<p>I feel like I learned a lot already. It’s been a dense week that I would like to sum up.</p>

<h1 id="sequences-representation">Sequences representation</h1>

<p>Learning bioinformatics implies learning first about a number of file formats. Each file type differ from the other in the way data is represented and it’s primordial to know the general structure of the file in order to manipulate it properly.
Here, we will cover:</p>
<ul>
<li> fasta </li>
<li> fastq </li>
<li> BED </li>
<li> gtf/gff </li>
<li> sam/bam </li>
</ul>

<h4 id="fasta">fasta</h4>
<p>A fasta file is a simple way of representing nucleotide and amino-acid sequences. It can store one or many records. Each record contains a one-line header that starts with a greater than sign ‘&gt;’and contains basic information about the sequence (accession number, ID, name, organism, you name it..) followed by the sequence that can contain one line or more.
It is important to note that any tabulators, spaces, asterisks etc in sequence will be ignored.
You can comment lines i.e ignore them using the ‘;’</p>

<figure class="highlight"><pre><code class="language-ruby" data-lang="ruby"><span class="o">&gt;</span><span class="n">gi</span><span class="o">|</span><span class="mi">9626243</span><span class="o">|</span><span class="n">ref</span><span class="o">|</span><span class="no">NC_001416</span><span class="o">.</span><span class="mi">1</span><span class="o">|</span> <span class="no">Enterobacteria</span> <span class="n">phage</span> <span class="nb">lambda</span><span class="p">,</span> <span class="n">complete</span> <span class="n">genome</span>
<span class="no">GGGCGGCGACCTCGCGGGTTTTCGCTATTTATGAAAATTTTCCGGTTTAAGGCGTTTCCGTTCTTCTTCG</span>
<span class="no">TCATAACTTAATGTTTTTATTTAAAATACCCTCTGAAAAGAAAGGAAACGACAGGTGCTGAAAGCGAGGC</span>
<span class="no">TTTTTGGCCTCTGTCGTTTCCTTTCTCTGTTTTTGTCCGTGGAATGAACAATGGAAGTCAACAAAAAGCA</span>
<span class="c1">##;Ignore me</span></code></pre></figure>

<h4 id="fastq">fastq</h4>
<p>With the advent of next generation sequencing technologies (NGS), we started adding more information to the sequences, which becomes important in bioinformatics applications that need to take into account the quality. For this purpose emerged the Fastq Format.
Typically, each record in this file type contains 4 lines:</p>
<ul>
  <li>A header that starts with the sign ‘@’</li>
  <li>The raw sequence.</li>
  <li>A line that start with ‘+’ and can contain additional information about the sequence.</li>
  <li>A line that  encodes the quality value for the sequence in line 2 using ASCII characters.</li>
</ul>

<figure class="highlight"><pre><code class="language-ruby" data-lang="ruby"><span class="vi">@SRR835775</span><span class="o">.</span><span class="mi">1</span> <span class="mi">1</span><span class="o">/</span><span class="mi">1</span>
<span class="no">TAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTCACCCTAACCCTAACCCTAACCGTATCCGTCACCCTAACCCTAAC</span>
<span class="o">+</span>
<span class="sc">??</span><span class="p">?</span><span class="no">B1ADDD8</span><span class="p">??</span><span class="no">BB</span><span class="o">+</span><span class="no">C</span><span class="p">?</span><span class="no">B</span><span class="o">+</span><span class="p">:</span><span class="no">AA883CEE8</span><span class="p">?</span><span class="no">C3</span><span class="vi">@DDD3</span><span class="p">)?</span><span class="no">D2</span><span class="p">;</span><span class="no">DC</span><span class="p">?</span><span class="mi">8</span><span class="p">?</span><span class="o">=</span><span class="no">BAD</span><span class="o">=</span><span class="vi">@C</span><span class="err">@</span><span class="p">(</span><span class="o">.</span><span class="mf">6.6</span><span class="o">=</span><span class="no">A</span><span class="p">?</span><span class="o">=</span><span class="sc">?@</span><span class="c1">##################################</span></code></pre></figure>

<h4 id="browser-extensible-data-bed">Browser Extensible Data (BED)</h4>
<p>BED (Browser Extensible Data) format provides a flexible way to define the data lines that are displayed in an annotation track. BED lines have three required fields (Scaffold: chromose, transcript…, the 0-based start, and the end coordinates of the sequences considered) and nine additional optional fields(name, score, strand…). The columns are separated by tabs or spaces and the number of fields per line must be consistent throughout any single set of data in an annotation track.
Because of their simplicity BED files are widely use, and we will see in another post how to retreive bedfiles and analyze them using bedtools.</p>

<figure class="highlight"><pre><code class="language-ruby" data-lang="ruby"><span class="n">chr22</span>   <span class="mi">10510227</span>        <span class="mi">10510528</span>        <span class="no">AluSx1</span>  <span class="mi">2021</span>    <span class="o">+</span>
<span class="n">chr22</span>   <span class="mi">10514778</span>        <span class="mi">10515050</span>        <span class="no">AluSx1</span>  <span class="mi">1933</span>    <span class="o">+</span>
<span class="n">chr22</span>   <span class="mi">10519673</span>        <span class="mi">10519746</span>        <span class="no">AluJo</span>   <span class="mi">474</span>     <span class="o">-</span>
<span class="n">chr22</span>   <span class="mi">10520950</span>        <span class="mi">10521193</span>        <span class="no">AluJo</span>   <span class="mi">1258</span>    <span class="o">-</span>
<span class="n">chr22</span>   <span class="mi">10522328</span>        <span class="mi">10522608</span>        <span class="no">AluSg</span>   <span class="mi">2274</span>    <span class="o">+</span>
<span class="n">chr22</span>   <span class="mi">10528506</span>        <span class="mi">10528771</span>        <span class="no">AluJr</span>   <span class="mi">1344</span>    <span class="o">+</span>
<span class="n">chr22</span>   <span class="mi">10529446</span>        <span class="mi">10529561</span>        <span class="no">AluJo</span>   <span class="mi">771</span>     <span class="o">+</span></code></pre></figure>

<h4 id="sequence-alignment-map-sam">Sequence Alignment Map (SAM)</h4>
<p>It is a TAB-delimited text format with a:</p>
<ul>
  <li>Multi-line header: every line starting with ‘@’. It contains informations about the number of segments, the number of sequences, the program by which it has been generation and if the sequences are sorted or not. Note that the header may be absent.</li>
</ul>

<figure class="highlight"><pre><code class="language-ruby" data-lang="ruby"><span class="vi">@HD</span>     <span class="no">VN</span><span class="p">:</span><span class="mf">1.0</span>  <span class="no">GO</span><span class="ss">:none</span> <span class="no">SO</span><span class="p">:</span> <span class="no">Sorted</span> <span class="n">by</span>
<span class="vi">@SQ</span>     <span class="no">SN</span><span class="ss">:Chr1</span> <span class="no">LN</span><span class="p">:</span><span class="mi">2992</span>   
<span class="vi">@SQ</span>     <span class="no">SN</span><span class="ss">:Chr2</span> <span class="no">LN</span><span class="p">:</span><span class="mi">19381</span> 
<span class="vi">@RG</span>     <span class="no">ID</span><span class="ss">:H100223_GAII05_0002</span>  <span class="no">PL</span><span class="ss">:SLX</span>  <span class="no">LB</span><span class="ss">:wu_PII</span>       <span class="no">PI</span><span class="p">:</span><span class="mi">400</span>  <span class="no">DS</span><span class="ss">:wu_0_Genome</span>  <span class="no">SM</span><span class="ss">:wu_0</span>
<span class="vi">@RG</span>     <span class="no">ID</span><span class="ss">:Wii_PER01</span>    <span class="no">PL</span><span class="ss">:SLX</span>  <span class="no">LB</span><span class="ss">:wu_phaseI</span>    <span class="no">PI</span><span class="p">:</span><span class="mi">400</span>  <span class="no">DS</span><span class="ss">:wu_0_Genome</span>  <span class="no">SM</span><span class="ss">:wu_0</span>
<span class="vi">@PG</span>     <span class="no">ID</span><span class="ss">:Program</span> <span class="nb">name</span>       <span class="no">VN</span><span class="p">:</span> <span class="n">program</span> <span class="n">version</span> <span class="no">CL</span><span class="p">:</span> <span class="no">Command</span> <span class="n">line</span>
<span class="vi">@CO</span>     <span class="no">TM</span><span class="ss">:Fri</span><span class="p">,</span> <span class="mi">17</span> <span class="no">Sep</span> <span class="mi">2010</span> <span class="mi">12</span><span class="p">:</span><span class="mi">20</span><span class="p">:</span><span class="mi">13</span> <span class="no">BST</span>        <span class="no">WD</span><span class="ss">:/</span><span class="n">lustre</span><span class="o">/</span><span class="n">scratch103</span><span class="o">/</span><span class="n">sanger</span><span class="o">/</span><span class="n">xcg</span><span class="o">/</span><span class="n">wu_0</span><span class="o">/</span><span class="nb">self</span>      <span class="no">HN</span><span class="ss">:bc</span><span class="o">-</span><span class="mi">16</span><span class="o">-</span><span class="mi">1</span><span class="o">-</span><span class="mo">07</span><span class="p">.</span><span class="nf">internal</span><span class="p">.</span><span class="nf">sanger</span><span class="p">.</span><span class="nf">ac</span><span class="p">.</span><span class="nf">uk</span>     <span class="no">UN</span><span class="ss">:xcg</span></code></pre></figure>

<ul>
  <li>body: constituted  by the alignment records. Each alignment line has 11 <strong>mandatory</strong> that always come in the same order and contain the information that follows:</li>
</ul>
<ol>
   <li>QNAME : Query template NAME. </li>
    <li>FLAG: combination of multiple series of information all of which are represented in binary (bits). </li>
    <li>RNAME:  Reference sequence NAME of the alignment it matches a @SQ field in the header section. </li>
     <li>POS : 1-based leftmost mapping position. </li>
    <li>MAPQ : Mapping quality. </li>
    <li>CIGAR : stands for Concise Idiosyncratic Gapped Alignment Report it describes the alignment (insertions, deletions, matches...). </li>
    <li>RNEXT : Reference sequence name of the primary alignment of the NEXT read in the template</li>
    <li>PNEXT: Position of the next read in the template. (1-based)</li>
    <li>TLEN:	Template length. </li>
    <li>SEQ:	Read Sequence. </li>
    <li>QUAL: Read Quality. </li>
  </ol>

<figure class="highlight"><pre><code class="language-ruby" data-lang="ruby"><span class="vi">@SRR835775</span><span class="o">.</span><span class="mi">1</span> <span class="mi">1</span><span class="o">/</span><span class="mi">1</span>
<span class="n">readID43GYAX15</span><span class="p">:</span><span class="mi">7</span><span class="p">:</span><span class="mi">1</span><span class="p">:</span><span class="mi">1202</span><span class="p">:</span><span class="mi">19894</span><span class="o">/</span><span class="mi">1</span>    <span class="mi">256</span>    <span class="n">contig87</span>    <span class="mi">540849</span>    <span class="mi">1</span>    <span class="mi">65</span><span class="no">M</span>    <span class="o">*</span>    <span class="mi">0</span>    <span class="mi">0</span>    <span class="no">CCTGCACGAACGAAATCCGCATGCGTCTGGTCGTTGTACGGAACGGCGGTTGTGTGACGAACGGC</span>    <span class="no">EDDEEDEE</span><span class="o">=</span><span class="no">EE</span><span class="p">?</span><span class="no">DE</span><span class="p">??</span><span class="no">DDDBADEBEFFFDBEFFEBCBC</span><span class="o">=</span><span class="p">?</span><span class="no">BEEEE</span><span class="err">@</span><span class="o">=</span><span class="p">:</span><span class="sc">?:</span><span class="p">:</span><span class="sc">?7</span><span class="p">?:</span><span class="mi">8</span><span class="o">-</span><span class="mi">6</span><span class="p">?</span><span class="mi">7</span><span class="p">?</span><span class="err">@</span><span class="p">?</span><span class="sc">?#</span> <span class="no">AS</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>  <span class="no">XS</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span> <span class="no">XN</span><span class="p">:</span><span class="n">i</span><span class="p">:</span><span class="mi">0</span>  <span class="no">XM</span><span class="ss">:i</span><span class="p">:</span><span class="mi">0</span> <span class="no">XO</span><span class="ss">:i</span><span class="p">:</span><span class="mi">0</span>  <span class="no">XG</span><span class="ss">:i</span><span class="p">:</span><span class="mi">0</span>  <span class="no">NM</span><span class="ss">:i</span><span class="p">:</span><span class="mi">0</span>    <span class="no">MD</span><span class="ss">:Z</span><span class="p">:</span><span class="mi">65</span>    <span class="no">YT</span><span class="ss">:Z:UU</span></code></pre></figure>

<p>You can read more about the SAM format <a href="http://samtools.github.io/hts-specs/SAMv1.pdf" target="_about"> here </a>.</p>

<h4 id="bam-binary-alignment-map">Bam (Binary Alignment Map)</h4>
<p>BAM files are compressed (binary) SAM files. They contain the same information.Their small size make them ideal to store alignment files. And that’s when the command line comes in handy. We will use samtools in order to view the file in the upcoming post.</p>

  </div>

</article>

      </div>
    </div>

    <footer class="site-footer">

  <div class="wrapper">

    <h2 class="footer-heading">Digital codon</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col footer-col-1">
        <ul class="contact-list">
          <li>Digital codon</li>
          <li><a href="mailto:benmeradi.lilyanezli@gmail.com">benmeradi.lilyanezli@gmail.com</a></li>
        </ul>
      </div>

      <div class="footer-col footer-col-2">
        <ul class="social-media-list">
          
          <li>
            <a href="https://github.com/nezlicodes"><span class="icon icon--github"><svg viewBox="0 0 16 16"><path fill="#828282" d="M7.999,0.431c-4.285,0-7.76,3.474-7.76,7.761 c0,3.428,2.223,6.337,5.307,7.363c0.388,0.071,0.53-0.168,0.53-0.374c0-0.184-0.007-0.672-0.01-1.32 c-2.159,0.469-2.614-1.04-2.614-1.04c-0.353-0.896-0.862-1.135-0.862-1.135c-0.705-0.481,0.053-0.472,0.053-0.472 c0.779,0.055,1.189,0.8,1.189,0.8c0.692,1.186,1.816,0.843,2.258,0.645c0.071-0.502,0.271-0.843,0.493-1.037 C4.86,11.425,3.049,10.76,3.049,7.786c0-0.847,0.302-1.54,0.799-2.082C3.768,5.507,3.501,4.718,3.924,3.65 c0,0,0.652-0.209,2.134,0.796C6.677,4.273,7.34,4.187,8,4.184c0.659,0.003,1.323,0.089,1.943,0.261 c1.482-1.004,2.132-0.796,2.132-0.796c0.423,1.068,0.157,1.857,0.077,2.054c0.497,0.542,0.798,1.235,0.798,2.082 c0,2.981-1.814,3.637-3.543,3.829c0.279,0.24,0.527,0.713,0.527,1.437c0,1.037-0.01,1.874-0.01,2.129 c0,0.208,0.14,0.449,0.534,0.373c3.081-1.028,5.302-3.935,5.302-7.362C15.76,3.906,12.285,0.431,7.999,0.431z"/></svg>
</span><span class="username">nezlicodes</span></a>

          </li>
          

          
          <li>
            <a href="https://twitter.com/nezlicodes"><span class="icon icon--twitter"><svg viewBox="0 0 16 16"><path fill="#828282" d="M15.969,3.058c-0.586,0.26-1.217,0.436-1.878,0.515c0.675-0.405,1.194-1.045,1.438-1.809c-0.632,0.375-1.332,0.647-2.076,0.793c-0.596-0.636-1.446-1.033-2.387-1.033c-1.806,0-3.27,1.464-3.27,3.27 c0,0.256,0.029,0.506,0.085,0.745C5.163,5.404,2.753,4.102,1.14,2.124C0.859,2.607,0.698,3.168,0.698,3.767 c0,1.134,0.577,2.135,1.455,2.722C1.616,6.472,1.112,6.325,0.671,6.08c0,0.014,0,0.027,0,0.041c0,1.584,1.127,2.906,2.623,3.206 C3.02,9.402,2.731,9.442,2.433,9.442c-0.211,0-0.416-0.021-0.615-0.059c0.416,1.299,1.624,2.245,3.055,2.271 c-1.119,0.877-2.529,1.4-4.061,1.4c-0.264,0-0.524-0.015-0.78-0.046c1.447,0.928,3.166,1.469,5.013,1.469 c6.015,0,9.304-4.983,9.304-9.304c0-0.142-0.003-0.283-0.009-0.423C14.976,4.29,15.531,3.714,15.969,3.058z"/></svg>
</span><span class="username">nezlicodes</span></a>

          </li>
          
        </ul>
      </div>

      <div class="footer-col footer-col-3">
        <p>Write an awesome description for your new site here. You can edit this line in _config.yml. It will appear in your document head meta (for Google search results) and in your feed.xml site description.
</p>
      </div>
    </div>

  </div>

</footer>


  </body>

</html>
